timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -430,4 +430,156 @@ void __global__ k_nonbonded_unified(
     }
 }
 
+template <typename RealType, int THREADS>
+void __global__ k_compute_nonbonded_target_atom_energies(
+    const int N,
+    const int num_target_atoms,
+    const int *__restrict__ target_atoms,        // [num_target_atoms]
+    const int *__restrict__ target_mols,         // [num_target_atoms]
+    const int *__restrict__ target_mols_offsets, // [num_mols + 1]
+    const double *__restrict__ coords,           // [N, 3]
+    const double *__restrict__ params,           // [N, PARAMS_PER_ATOM]
+    const double *__restrict__ box,              // [3, 3],
+    const RealType beta,
+    const RealType cutoff_squared,
+    __int128 *__restrict__ output_energies // [num_target_atoms, gridDim.x]
+) {
+    static_assert(THREADS <= 256 && (THREADS & (THREADS - 1)) == 0);
+    volatile __shared__ __int128 block_energy_buffer[THREADS];
+
+    int row_idx = blockIdx.y;
+    if (row_idx >= num_target_atoms) {
+        return;
+    }
+    const RealType bx = box[0 * 3 + 0];
+    const RealType by = box[1 * 3 + 1];
+    const RealType bz = box[2 * 3 + 2];
+
+    const RealType inv_bx = 1 / bx;
+    const RealType inv_by = 1 / by;
+    const RealType inv_bz = 1 / bz;
+
+    const int atom_i_idx = target_atoms[row_idx];
+    const int mol_i_idx = target_mols[row_idx];
+
+    const int min_mol_offset = target_mols_offsets[mol_i_idx];
+    const int max_mol_offset = target_mols_offsets[mol_i_idx + 1];
+    const int min_atom_idx = target_atoms[min_mol_offset];
+    const int max_atom_idx = target_atoms[max_mol_offset - 1];
+
+    int params_i_idx = atom_i_idx * PARAMS_PER_ATOM;
+    int charge_param_idx_i = params_i_idx + PARAM_OFFSET_CHARGE;
+    int lj_param_idx_sig_i = params_i_idx + PARAM_OFFSET_SIG;
+    int lj_param_idx_eps_i = params_i_idx + PARAM_OFFSET_EPS;
+    int w_param_idx_i = params_i_idx + PARAM_OFFSET_W;
+
+    RealType qi = params[charge_param_idx_i];
+    RealType sig_i = params[lj_param_idx_sig_i];
+    RealType eps_i = params[lj_param_idx_eps_i];
+    RealType w_i = params[w_param_idx_i];
+
+    RealType ci_x = coords[atom_i_idx * 3 + 0];
+    RealType ci_y = coords[atom_i_idx * 3 + 1];
+    RealType ci_z = coords[atom_i_idx * 3 + 2];
+
+    int atom_j_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    // Zero out the energy buffer
+    block_energy_buffer[threadIdx.x] = 0;
+    while (atom_j_idx < N) {
+        // The two atoms are in the same molecule, don't compute the energies
+        if (!(atom_j_idx >= min_atom_idx && atom_j_idx <= max_atom_idx)) {
```

Comment:
maybe simplify the predicate logic here?

Define predicates `X = atom_j_idx >= min_atom_idx`, `Y = atom_j_idx <= max_atom_idx`:

`!(X AND Y) == !X OR !Y == (atom_j_idx < min_atom_idx) OR atom_j_idx > max_atom_idx`

Reply:
Also maybe add a comment re: assumptions about continuity of atoms within a molecule.

Reply:
Fixed up in https://github.com/proteneer/timemachine/pull/1198/commits/39a594604e770acc32223537222e24a9b0324be3 and added comment about assumption